File: relay/internal/channel/noise.go
// Package channel implements the relay side of the carrier-independent
// device↔realm channel (ADR-008): a Noise IK session carrying CBOR frames,
// fragmented into Noise messages of at most 65 535 bytes. It mirrors
// arveil-core's `channel` module byte for byte; the Rust test vectors in
// codec_test.go are the contract.
package channel

import (
	"crypto/rand"
	"errors"
	"fmt"

	"github.com/flynn/noise"
)

// ProtocolName is fixed for V1 and shared with arveil-core.
const ProtocolName = "Noise_IK_25519_ChaChaPoly_BLAKE2s"

// ProtocolVersion is the wire protocol major version (see relay/internal/version).
const ProtocolVersion uint16 = 0

const (
	// MaxNoiseMessage is the Noise spec limit for one message.
	MaxNoiseMessage = noise.MaxMsgLen
	// MaxNoisePayload is the largest payload of one transport message.
	MaxNoisePayload = MaxNoiseMessage - 16
)

var (
	ErrPayloadTooLarge      = errors.New("noise: payload exceeds one message")
	ErrUnexpectedPayload    = errors.New("noise: handshake message carried unexpected payload")
	ErrHandshakeNotFinished = errors.New("noise: handshake not finished")
)

var suite = noise.NewCipherSuite(noise.DH25519, noise.CipherChaChaPoly, noise.HashBLAKE2s)

// Prologue binds protocol version and realm identity, exactly as arveil-core does.
func Prologue(realmID []byte) []byte {
	p := []byte(fmt.Sprintf("arveil/%d/", ProtocolVersion))
	return append(p, realmID...)
}

// GenerateStaticKeypair returns a fresh X25519 keypair for the channel.
func GenerateStaticKeypair() (noise.DHKey, error) {
	return suite.GenerateKeypair(rand.Reader)
}

// StaticKeypairFromPrivate rebuilds a keypair from a stored 32-byte private key.
func StaticKeypairFromPrivate(private []byte) (noise.DHKey, error) {
	if len(private) != 32 {
		return noise.DHKey{}, fmt.Errorf("noise: private key must be 32 bytes, got %d", len(private))
	}
	// X25519 public key derivation via the suite's DH with the base point is
	// not exposed by flynn/noise; use x/crypto's curve25519 through DH25519.
	pub, err := x25519Public(private)
	if err != nil {
		return noise.DHKey{}, err
	}
	return noise.DHKey{Private: append([]byte(nil), private...), Public: pub}, nil
}

// Responder is the realm side of the IK handshake.
type Responder struct {
	hs *noise.HandshakeState
}

// NewResponder prepares to answer one initiator.
func NewResponder(static noise.DHKey, prologue []byte) (*Responder, error) {
	hs, err := noise.NewHandshakeState(noise.Config{
		CipherSuite:   suite,
		Pattern:       noise.HandshakeIK,
		Initiator:     false,
		Prologue:      prologue,
		StaticKeypair: static,
	})
	if err != nil {
		return nil, err
	}
	return &Responder{hs: hs}, nil
}

// ReadMessage1 consumes the initiator's first message and returns its static
// public key. Nothing is authenticated for the initiator yet: the caller
// decides whether to answer at all, and processes no frame until a transport
// message opens.
func (r *Responder) ReadMessage1(message []byte) (initiatorStatic []byte, err error) {
	payload, _, _, err := r.hs.ReadMessage(nil, message)
	if err != nil {
		return nil, err
	}
	if len(payload) != 0 {
		return nil, ErrUnexpectedPayload
	}
	return append([]byte(nil), r.hs.PeerStatic()...), nil
}

// WriteMessage2 produces the responder's message and the established transport.
func (r *Responder) WriteMessage2() (message []byte, t *Transport, err error) {
	// flynn/noise returns, once the handshake completes, the CipherState for
	// initiator→responder first and responder→initiator second.
	out, csI2R, csR2I, err := r.hs.WriteMessage(nil, nil)
	if err != nil {
		return nil, nil, err
	}
	if csI2R == nil || csR2I == nil {
		return nil, nil, ErrHandshakeNotFinished
	}
	return out, &Transport{
		enc:          csR2I,
		dec:          csI2R,
		remoteStatic: append([]byte(nil), r.hs.PeerStatic()...),
	}, nil
}

// Initiator is the device side; the relay only needs it for tests and
// for talking to other relays in the future.
type Initiator struct {
	hs *noise.HandshakeState
}

func NewInitiator(static noise.DHKey, remoteStatic, prologue []byte) (*Initiator, error) {
	hs, err := noise.NewHandshakeState(noise.Config{
		CipherSuite:   suite,
		Pattern:       noise.HandshakeIK,
		Initiator:     true,
		Prologue:      prologue,
		StaticKeypair: static,
		PeerStatic:    remoteStatic,
	})
	if err != nil {
		return nil, err
	}
	return &Initiator{hs: hs}, nil
}

func (i *Initiator) WriteMessage1() ([]byte, error) {
	out, _, _, err := i.hs.WriteMessage(nil, nil)
	return out, err
}

func (i *Initiator) ReadMessage2(message []byte) (*Transport, error) {
	payload, csI2R, csR2I, err := i.hs.ReadMessage(nil, message)
	if err != nil {
		return nil, err
	}
	if len(payload) != 0 {
		return nil, ErrUnexpectedPayload
	}
	if csI2R == nil || csR2I == nil {
		return nil, ErrHandshakeNotFinished
	}
	return &Transport{
		enc:          csI2R,
		dec:          csR2I,
		remoteStatic: append([]byte(nil), i.hs.PeerStatic()...),
	}, nil
}

// Transport is an established session. One Seal produces one wire message.
type Transport struct {
	enc, dec     *noise.CipherState
	remoteStatic []byte
}

// RemoteStatic is the peer's static public key authenticated by the handshake.
func (t *Transport) RemoteStatic() []byte { return t.remoteStatic }

func (t *Transport) Seal(payload []byte) ([]byte, error) {
	if len(payload) > MaxNoisePayload {
		return nil, ErrPayloadTooLarge
	}
	return t.enc.Encrypt(nil, nil, payload)
}

func (t *Transport) Open(message []byte) ([]byte, error) {
	if len(message) > MaxNoiseMessage {
		return nil, ErrPayloadTooLarge
	}
	return t.dec.Decrypt(nil, nil, message)
}

File: core/crates/arveil-core/src/channel/noise.rs
//! Noise `IK` handshake and transport over `snow`.
//!
//! The initiator (device) knows the responder's (realm's) static key from
//! the bootstrap QR or the signed endpoint list. The responder learns the
//! initiator's static key from the handshake and must check it against a
//! device credential before serving any frame. The first message carries no
//! application payload (ADR-008: `IK` gives it neither forward secrecy nor
//! replay protection).

use snow::params::NoiseParams;
use snow::{Builder, HandshakeState, TransportState};

use crate::PROTOCOL_VERSION;

/// Fixed for V1. Both `snow` and Go's `flynn/noise` implement it.
pub const PROTOCOL_NAME: &str = "Noise_IK_25519_ChaChaPoly_BLAKE2s";

/// Largest Noise message on the wire (Noise spec limit).
pub const MAX_NOISE_MESSAGE: usize = 65_535;
/// Largest payload that fits in one Noise transport message.
pub const MAX_NOISE_PAYLOAD: usize = MAX_NOISE_MESSAGE - 16;

#[derive(Debug, thiserror::Error)]
pub enum NoiseError {
    #[error("noise: {0}")]
    Snow(#[from] snow::Error),
    #[error("noise: payload of {0} bytes exceeds one message")]
    PayloadTooLarge(usize),
    #[error("noise: handshake message carried unexpected payload")]
    UnexpectedHandshakePayload,
}

/// Prologue binding the protocol version and the realm identity, so a
/// handshake completed against the wrong realm or version fails even when
/// the keys match.
pub fn prologue(realm_id: &[u8]) -> Vec<u8> {
    let mut p = format!("arveil/{PROTOCOL_VERSION}/").into_bytes();
    p.extend_from_slice(realm_id);
    p
}

fn params() -> NoiseParams {
    PROTOCOL_NAME.parse().expect("valid Noise protocol name")
}

/// A static X25519 keypair for the Noise channel.
#[derive(Clone)]
pub struct StaticKeypair {
    pub private: Vec<u8>,
    pub public: Vec<u8>,
}

impl StaticKeypair {
    pub fn generate() -> Result<Self, NoiseError> {
        let kp = Builder::new(params()).generate_keypair()?;
        Ok(Self {
            private: kp.private,
            public: kp.public,
        })
    }
}

/// Device side of the handshake.
pub struct Initiator {
    state: HandshakeState,
}

impl Initiator {
    pub fn new(
        local: &StaticKeypair,
        remote_static_public: &[u8],
        prologue: &[u8],
    ) -> Result<Self, NoiseError> {
        let state = Builder::new(params())
            .local_private_key(&local.private)?
            .remote_public_key(remote_static_public)?
            .prologue(prologue)?
            .build_initiator()?;
        Ok(Self { state })
    }

    /// Message 1, with an empty payload by design.
    pub fn write_message_1(&mut self) -> Result<Vec<u8>, NoiseError> {
        let mut buf = vec![0u8; MAX_NOISE_MESSAGE];
        let n = self.state.write_message(&[], &mut buf)?;
        buf.truncate(n);
        Ok(buf)
    }

    /// Message 2 from the responder; on success the channel is established.
    pub fn read_message_2(mut self, message: &[u8]) -> Result<Transport, NoiseError> {
        let mut buf = vec![0u8; MAX_NOISE_MESSAGE];
        let n = self.state.read_message(message, &mut buf)?;
        if n != 0 {
            return Err(NoiseError::UnexpectedHandshakePayload);
        }
        Ok(Transport {
            state: self.state.into_transport_mode()?,
        })
    }
}

/// Realm side of the handshake.
pub struct Responder {
    state: HandshakeState,
}

impl Responder {
    pub fn new(local: &StaticKeypair, prologue: &[u8]) -> Result<Self, NoiseError> {
        let state = Builder::new(params())
            .local_private_key(&local.private)?
            .prologue(prologue)?
            .build_responder()?;
        Ok(Self { state })
    }

    /// Message 1 from the initiator. Returns the initiator's static public
    /// key so the caller can decide whether to answer at all.
    pub fn read_message_1(&mut self, message: &[u8]) -> Result<Vec<u8>, NoiseError> {
        let mut buf = vec![0u8; MAX_NOISE_MESSAGE];
        let n = self.state.read_message(message, &mut buf)?;
        if n != 0 {
            return Err(NoiseError::UnexpectedHandshakePayload);
        }
        Ok(self
            .state
            .get_remote_static()
            .expect("IK message 1 carries the initiator static")
            .to_vec())
    }

    /// Message 2; on success the channel is established.
    pub fn write_message_2(mut self) -> Result<(Vec<u8>, Transport), NoiseError> {
        let mut buf = vec![0u8; MAX_NOISE_MESSAGE];
        let n = self.state.write_message(&[], &mut buf)?;
        buf.truncate(n);
        Ok((
            buf,
            Transport {
                state: self.state.into_transport_mode()?,
            },
        ))
    }
}

/// Established Noise transport. One `seal` produces one wire message.
pub struct Transport {
    state: TransportState,
}

impl Transport {
    pub fn remote_static(&self) -> &[u8] {
        self.state
            .get_remote_static()
            .expect("IK authenticates the remote static")
    }

    pub fn seal(&mut self, payload: &[u8]) -> Result<Vec<u8>, NoiseError> {
        if payload.len() > MAX_NOISE_PAYLOAD {
            return Err(NoiseError::PayloadTooLarge(payload.len()));
        }
        let mut buf = vec![0u8; MAX_NOISE_MESSAGE];
        let n = self.state.write_message(payload, &mut buf)?;
        buf.truncate(n);
        Ok(buf)
    }

    pub fn open(&mut self, message: &[u8]) -> Result<Vec<u8>, NoiseError> {
        if message.len() > MAX_NOISE_MESSAGE {
            return Err(NoiseError::PayloadTooLarge(message.len()));
        }
        let mut buf = vec![0u8; MAX_NOISE_MESSAGE];
        let n = self.state.read_message(message, &mut buf)?;
        buf.truncate(n);
        Ok(buf)
    }
}
